crawlers/mooncrawl/eth_scan.py

```diff
@@ -0,0 +1,184 @@
+    raise ValueError("AWS_S3_SMARTCONTRACT_BUCKET must be set")
+
+
+@dataclass
+class VerifiedSmartContract:
+    name: str
+    address: str
+    tx_hash: str
+
+
+def push_to_bucket(contract_data: Dict[str, Any], contract_file: str):
+    result_bytes = json.dumps(contract_data).encode("utf-8")
+    result_key = contract_file
+
+    s3 = boto3.client("s3")
+    s3.put_object(
+        Body=result_bytes,
+        Bucket=bucket,
+        Key=result_key,
+        ContentType="application/json",
+        Metadata={"source": "etherscan", "crawler_version": ""},
+    )
+
+
+def get_address_id(db_session: Session, contract_address: str) -> int:
+    """
+    Searches for given address in EthereumAddress table,
+    If doesn't find one, creates new.
+    Returns id of address
+    """
+    query = db_session.query(EthereumAddress.id).filter(
+        EthereumAddress.address == contract_address
+    )
+    id = query.one_or_none()
+    if id is not None:
+        return id[0]
+
+    latest_address_id = (
+        db_session.query(EthereumAddress.id).order_by(text("id desc")).limit(1).one()
+    )[0]
+
+    id = latest_address_id + 1
+    smart_contract = EthereumAddress(
+        id=id,
+        address=contract_address,
+    )
+    db_session.add(smart_contract)
+    db_session.commit()
+    return id
+
+
+def crawl_step(db_session: Session, contract: VerifiedSmartContract, crawl_url: str):
+    attempt = 0
+    current_interval = 2
+    success = False
+
+    response: Optional[requests.Response] = None
+    while (not success) and attempt < 3:
+        attempt += 1
+        try:
+            response = requests.get(crawl_url)
+            response.raise_for_status()
+            success = True
+        except:
+            current_interval *= 2
+            time.sleep(current_interval)
+
+    if response is None:
+        print(f"Could not process URL: {crawl_url}", file=sys.stderr)
+        return None
+    page = response.json()
+    result = page["result"][0]
+    contract_info = {
+        "data": result,
+        "crawl_version": MOONCRAWL_VERSION,
+        "crawled_at": f"{datetime.now()}",
+    }
+    object_key = f"/v1/etherscan/{contract.address}.json"
+    push_to_bucket(contract_info, object_key)
+
+    eth_address_id = get_address_id(db_session, contract.address)
+
+    eth_label = EthereumLabel(
+        label=ETHERSCAN_SMARTCONTRACTS_LABEL_NAME,
+        address_id=eth_address_id,
+        label_data={
+            "key": object_key,
+            "name": contract.name,
+            "tx_hash": contract.tx_hash,
+        },
+    )
+
+    db_session.add(eth_label)
+    db_session.commit()
+
+
+def crawl(
+    smart_contracts: List[VerifiedSmartContract],
+    interval: float,
+    start_step=0,
+):
+    with yield_db_session_ctx() as db_session:
+        for i in range(start_step, len(smart_contracts)):
+            contract = smart_contracts[i]
+            print(f"Crawling {i+1}/{len(smart_contracts)} : {contract.address}")
```

Comment:
You could end up with `"Crawling 2001/1000: <address>"`, right?

Maybe better message here.